### PR TITLE
Clone document on addStatic.

### PR DESCRIPTION
**Pull request by aljones15**
This seems like the easiest solution to the document mutation issue.
If we don't want a dep we could use `JSON.parse(JSON.stringify(document))` as the jsonld doc should be just JSON.

### CODE REVIEW

File: lib/JsonLdDocumentLoader.js [hunk 1]
```diff
@@ -15,7 +16,7 @@ export class JsonLdDocumentLoader {
     if(!_isObject(document)) {
       throw new TypeError('The second parameter (document) must be an object.');
     }
-    this.documents.set(url, document);
+    this.documents.set(url, klona(document));
```
Comment (dlongley) [inline lib/JsonLdDocumentLoader.js]: Would `structuredClone` work instead?

https://developer.mozilla.org/en-US/docs/Web/API/structuredClone

```suggestion
    this.documents.set(url, structuredClone(document));
```

**Comment by aljones15** [comment]
@dlongley `structuredClone` works, but not in node <= 16 came in to node in 17 according to the page you linked to.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,10 @@
 # jsonld-document-loader
 
+## 2.0.1 -
+
+### Changed
+- addStatic deep clones documents to avoid mutation.
```
Comment (dlongley) [inline CHANGELOG.md]:
```suggestion
- `addStatic` deep clones documents to avoid mutation.
```

File: test/10-test.spec.js [hunk 3]
```diff
@@ -96,6 +96,42 @@ describe('jsonld-document-loader', () => {
...
+      );
+      should.exist(result.document, 'Expected result to have a document.');
+      result.document.should.not.eql(
+        originalDoc,
+        'Expected document from documentLoader to not equal mutated document.'
+      );
+      result.document.should.eql(
+        expectedDoc,
+        'Expected document from documentLoader to equal unmutated document.'
+      );
```
Comment (dlongley) [inline test/10-test.spec.js]: DB style is to not drop the last parenthetical (unless preceded by `}`).

```suggestion
      should.exist(
        result,
        `Expected documentLoader to return a document for ${documentUrl}.`);
      result.should.be.an(
        'object',
        'Expected document to be an object.');
      should.exist(result.document, 'Expected result to have a document.');
      result.document.should.not.eql(
        originalDoc,
        'Expected document from documentLoader to not equal mutated document.');
      result.document.should.eql(
        expectedDoc,
        'Expected document from documentLoader to equal unmutated document.');
```

**Review by dlongley — APPROVED** [review]
Thanks!

**Comment by dlongley** [comment]
We should drop the testing for node 16 and bump the suggestion in the package.json (if any) to node 18+.

**Comment by aljones15** [comment]
@dlongley while bumping engines we had previously agreed is not major is adding a new method that is node >= 17 only and only supported in somewhat recent browsers breaking and hence a major change?

**Comment by dlongley** [comment]
Given our package.json "node recommendation" policy, it's not a breaking change. We've just gotten behind on upgrading this particular package's expected node value here. We only support LTS+. So we could do the `node >= 18` bump in a separate PR and release first, but it would still be a non-breaking release. Since there's no real difference there other than churn, let's roll this all together.

The supported browsers and node versions are "current" and LTS+, and think we're good there for the `structuredClone` feature:

https://caniuse.com/?search=structuredClone

If people need to polyfill this in old browsers I imagine they're already doing so in their build pipelines or should be. As a matter of policy, our individual libraries do not provide support for legacy platforms.

**Comment by davidlehn** [comment]
- Just to check, the speed and memory hit on this is ok?  I imagine in most use cases no one one notice.  But there's also no choice for those rare use cases with huge resources.  "Use another loader" is a choice of course.
- "engine" spec issues aside, the definite newer API usage here could break user code running in 16.x (for whatever legacy reasons that is being used).  Are you all sure you want to do that?

**Comment by davidlehn** [comment]
> * "engine" spec issues aside, the definite newer API usage here could break user code running in 16.x (for whatever legacy reasons that is being used).  Are you all sure you want to do that?

At a minimum, I think the changelog should note support or a polyfill for `structuredClone` is required and mention whatever version of Node.js that is.

**Review by davidlehn — APPROVED** [review]
I'm not convinced this is needed, but we'll try it out, and adjust the API if needed.

**Pull request merged**